networking/kvm.go

```diff
@@ -113,32 +220,87 @@ func kvmSetup(podRoot string, podID types.UUID, fps []ForwardedPort, privateNetL
 			}
 
 			// add address to host tap device
-			err = netlink.AddrAdd(
+			err = ensureHasAddr(
 				link,
-				&netlink.Addr{
-					IPNet: &net.IPNet{
-						IP:   n.runtime.HostIP,
-						Mask: net.IPMask(n.runtime.Mask),
-					},
-					Label: ifName,
-				})
+				&net.IPNet{
+					IP:   n.runtime.IP4.Gateway,
+					Mask: net.IPMask(n.runtime.Mask),
+				},
+			)
 			if err != nil {
 				return nil, fmt.Errorf("cannot add address to host tap device %q: %v", ifName, err)
 			}
 
-			if n.conf.IPMasq {
-				h := sha512.Sum512([]byte(podID.String()))
-				chain := fmt.Sprintf("CNI-%s-%x", n.conf.Name, h[:8])
-				if err = ip.SetupIPMasq(&net.IPNet{
-					IP:   n.runtime.IP,
-					Mask: net.IPMask(n.runtime.Mask),
-				}, chain); err != nil {
-					return nil, err
+			time.Sleep(time.Second / 2) // HACKHACKHACK - without this delay below changes in routes are not persistent
+			if err := removeAllRoutesOnLink(link); err != nil {
+				return nil, fmt.Errorf("cannot remove route on host tap device %q: %v", ifName, err)
+			}
+
+			if err := addRoute(link, n.runtime.IP); err != nil {
+				return nil, fmt.Errorf("cannot add on host direct route to pod: %v", err)
+			}
+
+		case "bridge":
+			config := BridgeNetConf{
+				NetConf: NetConf{
+					MTU: defaultMTU,
+				},
+				BrName: defaultBrName,
+			}
+			if err := json.Unmarshal(n.confBytes, &config); err != nil {
+				return nil, fmt.Errorf("error parsing %q result: %v", n.conf.Name, err)
+			}
+
+			br, err := ensureBridgeIsUp(config.BrName, config.MTU)
+			if err != nil {
+				return nil, fmt.Errorf("error in time of bridge setup: %v", err)
+			}
+			link, err := setupTapDevice(podID)
+			if err != nil {
+				return nil, fmt.Errorf("can not setup tap device: %v", err)
+			}
+			err = netlink.LinkSetMaster(link, br)
+			if err != nil {
+				// TODO: cleanup tap interface
+				return nil, fmt.Errorf("can not add tap interface to bridge: %v", err)
+			}
+
+			ifName := link.Attrs().Name
+			n.runtime.IfName = ifName
+
+			err = kvmSetupNetAddressing(&network, n, ifName)
+			if err != nil {
+				return nil, err
+			}
+
+			if config.IsGw {
+				err = ensureHasAddr(
+					br,
+					&net.IPNet{
+						IP:   n.runtime.IP4.Gateway,
+						Mask: net.IPMask(n.runtime.Mask),
+					},
+				)
+
+				if err != nil {
+					return nil, fmt.Errorf("cannot add address to host bridge device %q: %v", br.Name, err)
 				}
 			}
+
 		default:
 			return nil, fmt.Errorf("network %q have unsupported type: %q", n.conf.Name, n.conf.Type)
 		}
+
+		if n.conf.IPMasq {
+			h := sha512.Sum512([]byte(podID.String()))
+			chain := fmt.Sprintf("CNI-%s-%x", n.conf.Name, h[:8])
```

Comment:
these two lines "name of chain" is copy line 338-339 - can you consider it extracting it to common function eg. `podChain(podId, networkName string) string`